src/huggingface_hub/utils/cache_explorer.py

```diff
@@ -0,0 +1,246 @@
+
+    Any unexpected behavior will raise a `CorruptedCacheException`.
+    """
+    if not repo_path.is_dir():
+        raise CorruptedCacheException("Not a directory !")
+
+    if "--" not in repo_path.name:
+        raise CorruptedCacheException("Not a valid HuggingFace cache directory !")
+
+    repo_type, repo_id = repo_path.name.split("--", maxsplit=1)
+    if repo_type not in {"datasets", "models"}:
+        raise CorruptedCacheException("Repo type must be `dataset` or `model` !")
+    repo_type = repo_type[:-1]  # "models" -> "model"
+
+    blob_sizes: Dict[Path, int] = {}  # Key is blob_path, value is blob size (in bytes)
+
+    snapshots_path = repo_path / "snapshots"
+    refs_path = repo_path / "refs"
+    blobs_path = repo_path / "blobs"
+
+    if not snapshots_path.exists() or not snapshots_path.is_dir():
+        raise CorruptedCacheException("Snapshots dir doesn't exist !")
+
+    # Scan snapshots directory
+    cached_revisions_by_hash: Dict[str, CachedRevisionInfo] = {}
+    for revision_path in snapshots_path.iterdir():
+        if revision_path.is_file():
+            raise CorruptedCacheException("Snapshots folder corrupted: found a file  !")
+
+        cached_files = []
+        for file_path in revision_path.glob("**/*"):
+            # glob("**/*") iterates over all files and directories -> skip directories
+            if file_path.is_dir():
+                continue
+
+            if not file_path.is_symlink():
+                raise CorruptedCacheException(
+                    "Revision folder corrupted: found a non-symlink file !"
+                )
+
+            blob_path = Path(file_path).resolve()
+            if not blob_path.exists():
+                raise CorruptedCacheException("Blob missing (broken symlink) !")
+
+            if blobs_path not in blob_path.parents:
+                raise CorruptedCacheException(
+                    "Blob symlink points outside of blob directory !"
+                )
+
+            if blob_path not in blob_sizes:
+                blob_sizes[blob_path] = blob_path.stat().st_size
+
+            cached_files.append(
+                CachedFileInfo(
+                    file_name=file_path.name,
+                    file_path=file_path,
+                    size_on_disk=blob_sizes[blob_path],
+                    blob_path=blob_path,
+                )
+            )
+
+        cached_revisions_by_hash[revision_path.name] = CachedRevisionInfo(
+            commit_hash=revision_path.name,
+            size_on_disk=sum(
+                blob_sizes[blob_path]
+                for blob_path in set(file.blob_path for file in cached_files)
+            ),
+            files=sorted(cached_files, key=lambda file: file.file_path),
+        )
+
+    # Scan over `refs` directory
+    if refs_path.exists():
+        # Example of `refs` directory
+        # ── refs
+        #     ├── main
+        #     └── refs
+        #         └── pr
+        #             └── 1
```

Comment:
lol, I actually hadn't realized this file structure could be nested given that we allow `/` inside branches or tag names, @LysandreJik 

I'm not 100% sure this is robust though (on Windows for instance...) but i don't think it's a big deal either way

Reply:
Yep, discovered it quite randomly while working on that. I'd say it's robust enough because we would have known by now (🙄)

Reply:
The issue is if someone copy-pastes a cache folder from a Windows machine to a Unix-based one or the opposite.
We can always fix this in a separate PR but it seems a quite niche problem no ?